pkg/scheduler/objects/queue.go

```diff
@@ -1292,3 +1303,42 @@ func (sq *Queue) String() string {
 	return fmt.Sprintf("{QueuePath: %s, State: %s, StateTime: %x, MaxResource: %s}",
 		sq.QueuePath, sq.stateMachine.Current(), sq.stateTime, sq.maxResource)
 }
+
+func (sq *Queue) incRunningApps() {
+	sq.runningApps++
+	if sq.parent != nil {
+		sq.parent.incRunningApps()
+	}
+}
+
+func (sq *Queue) decRunningApps() {
+	sq.runningApps--
+	if sq.parent != nil {
+		sq.parent.decRunningApps()
+	}
+}
+
+func (sq *Queue) canRun() bool {
+	sq.RLock()
+	defer sq.RUnlock()
+	if sq.maxRunningApps == 0 {
+		return true
+	}
+	ok := sq.runningApps <= sq.maxRunningApps
+	if sq.parent != nil {
+		return ok && sq.parent.canRun()
+	}
+	return ok
+}
+
+func (sq *Queue) GetRunningApps() uint64 {
+	sq.RLock()
+	defer sq.RUnlock()
+	return sq.runningApps
+}
+
+func (sq *Queue) SetMaxRunningApps(maxApps uint64) {
+	sq.RLock()
+	defer sq.RUnlock()
```

Comment:
This is a read lock securing a write action...
Also please comment that this is test only as it is not called from anywhere else.